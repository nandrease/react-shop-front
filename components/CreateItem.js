import React, { Component } from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import Router from 'next/router';
import Form from './styles/Form';
import formatCurrency from '../lib/formatMoney';
import Error from './ErrorMessage';

const CREATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION(
		$title: String!
		$description: String!
		$price: Int!
		$image: String
		$largeImage: String
	) {
		createItem(title: $title, description: $description, price: $price, image: $image, largeImage: $largeImage) {
			id
		}
	}
`;

class CreateItem extends Component {
	state = {
		title: 'Cool shoes',
		description: 'Beautiful shoes',
		image: '',
		largeImage: '',
		price: 3400
	};
	handleChange = (e) => {
		const { name, type, value } = e.target;
		const val = type === 'number' && value ? parseFloat(value) : value;
		this.setState({ [name]: val });
	};

	uploadFile = async (e) => {
		const files = e.target.files;
		const data = new FormData();
		data.append('file', files[0]);
		data.append('upload_preset', 'shopreact');

		const res = await fetch('https://api.cloudinary.com/v1_1/doutr1bqe/image/upload', {
			method: 'POST',
			body: data
		});
		const file = await res.json();
		this.setState({
			image: file.secure_url,
			largeImage: file.eager[0].secure_url
		});
	};
	render() {
		return (
			<Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
				{(createItem, { loading, error }) => (
					<Form
						onSubmit={async (e) => {
							// stop form from submitting
							e.preventDefault();
							// call the mutation
							const res = await createItem();
							// change them to single item page
							Router.push({
								pathname: '/item',
								query: { id: res.data.createItem.id }
							});
						}}
					>
						<Error error={error} />
						<fieldset disabled={loading} aria-busy={loading}>
							<label htmlFor="file">
								Image
								<input
									type="file"
									id="file"
									name="file"
									placeholder="Upload an image"
									required
									onChange={this.uploadFile}
								/>
								{this.state.image && <img src={this.state.image} alt="Upload preview" width="200" />}
							</label>
							<label htmlFor="title">
								Title
								<input
									type="text"
									id="title"
									name="title"
									placeholder="Title"
									value={this.state.title}
									onChange={this.handleChange}
									required
								/>
							</label>
							<label htmlFor="price">
								Price
								<input
									type="number"
									id="price"
									name="price"
									placeholder="Price"
									value={this.state.price}
									onChange={this.handleChange}
									required
								/>
							</label>
							<label htmlFor="description">
								Description
								<textarea
									type="text"
									id="description"
									name="description"
									placeholder="Enter a description"
									value={this.state.description}
									onChange={this.handleChange}
									required
								/>
							</label>
							<button type="submit">Submit</button>
						</fieldset>
					</Form>
				)}
			</Mutation>
		);
	}
}

export default CreateItem;

export { CREATE_ITEM_MUTATION };
