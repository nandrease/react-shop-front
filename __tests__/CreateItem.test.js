import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import { MockedProvider } from 'react-apollo/test-utils';
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem';
import { fakeItem } from '../lib/testUtils';

const dogImage = 'https://url.com/dogimage.png';
// mock the global fetch API
global.fetch = jest.fn().mockResolvedValue({
	json: () => ({
		secure_url: dogImage,
		eager: [ { secure_url: dogImage } ]
	})
});

describe('<CreateItem />', () => {
	it('renders and matches screenshot', async () => {
		const wrapper = mount(
			<MockedProvider>
				<CreateItem />
			</MockedProvider>
		);
		const form = wrapper.find('form[data-test="form"]');
		expect(toJSON('form')).toMatchSnapshot();
	});

	it('uploads a file when changed', async () => {
		const wrapper = mount(
			<MockedProvider>
				<CreateItem />
			</MockedProvider>
		);
		const upload = wrapper.find('input[type="file"]');
		upload.simulate('change', { target: { files: [ 'fakedog.jpg' ] } });
		await wait();
		const component = wrapper.find('CreateItem').instance();
		expect(component.state.image).toEqual(dogImage);
		expect(component.state.largeImage).toEqual(dogImage);
		expect(global.fetch).toHaveBeenCalled();
		//expect(global.fetch).toHaveBeenCalledWith(
		// "https://api.cloudinary.com/v1_1/doutr1bqe/image/upload",
		// '{"body": {Symbol(impl): {"_entries": [{"name": "file", "value": "fakedog.jpg"}, {"name": "upload_preset", "value": "shopreact"}], Symbol(wrapper): [Circular]}}, "method": "POST"}'
		// );
		global.fetch.mockReset();
	});
	it('handles state updating', async () => {
		const wrapper = mount(
			<MockedProvider>
				<CreateItem />
			</MockedProvider>
		);
		wrapper.find('#title').simulate('change', {
			target: { value: 'Testing', name: 'title' }
		});
		wrapper.find('#price').simulate('change', {
			target: { value: 5000, name: 'price', type: 'number' }
		});
		wrapper.find('#description').simulate('change', {
			target: { value: 'This is description', name: 'description' }
		});

		expect(wrapper.find('CreateItem').instance().state).toMatchObject({
			title: 'Testing',
			price: 5000,
			description: 'This is description'
		});
	});
	it('creates an items when the form is submitted', async () => {
		const item = fakeItem();
		const mocks = [
			{
				request: {
					query: CREATE_ITEM_MUTATION,
					variables: {
						title: item.title,
						description: item.description,
						image: '',
						largeImage: '',
						price: item.price
					}
				},
				result: {
					data: {
						createItem: {
							...fakeItem,
							id: 'abc123',
							__typename: 'Item'
						}
					}
				}
			}
		];
		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<CreateItem />
			</MockedProvider>
		);
		// simulate filling out the form
		wrapper.find('#title').simulate('change', {
			target: { value: item.title, name: 'title' }
		});
		wrapper.find('#price').simulate('change', {
			target: { value: item.price, name: 'price', type: 'number' }
		});
		wrapper.find('#description').simulate('change', {
			target: { value: item.description, name: 'description' }
		});
		Router.router = { push: jest.fn() };
		wrapper.find('form[data-test="form"]').simulate('submit');
		await wait(50);
		expect(Router.router.push).toHaveBeenCalled();
		expect(Router.router.push).toHaveBeenCalledWith({"pathname": "/item", "query": {"id": "abc123"}});
	});
});
