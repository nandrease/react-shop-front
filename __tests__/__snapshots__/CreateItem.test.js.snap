// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem /> renders and matches screenshot 1`] = `undefined`;
